File: src/invoice/peppol.js
/**
 * bukio-cli — agent-first double-entry bookkeeping for SMEs across thirty-one jurisdictions.
 * Copyright (c) 2026 Erik van Kempen.
 * SPDX-License-Identifier: Apache-2.0
 */

// Peppol send (FR3.8): POST the UBL 2.1 / Peppol BIS 3.0 document to an
// access-point provider (e.g. Storecove-class). The provider contract is
// deliberately thin: POST the XML with a bearer token, expect 2xx.
// Configuration comes from the environment — never hard-code credentials:
//   BUKIO_PEPPOL_ENDPOINT  e.g. https://api.storecove.com/api/v2/invoices
//   BUKIO_PEPPOL_TOKEN     the provider API token

import { invoiceToUbl } from './ubl.js';

export function peppolError(code, message) {
  const e = new Error(message);
  e.code = code;
  return e;
}

export function peppolConfig() {
  return {
    endpoint: process.env.BUKIO_PEPPOL_ENDPOINT ?? null,
    token: process.env.BUKIO_PEPPOL_TOKEN ?? null,
  };
}

/**
 * Send a finalized invoice to the Peppol access point. dryRun validates the
 * configuration + payload without making the request.
 */
export async function sendPeppolInvoice(db, invoice, { endpoint = null, dryRun = false } = {}) {
  const cfg = peppolConfig();
  const ep = endpoint ?? cfg.endpoint;
  if (!ep) {
    throw peppolError('PEPPOL_NOT_CONFIGURED', 'no Peppol endpoint — set BUKIO_PEPPOL_ENDPOINT (or pass --endpoint)');
  }
  // Peppol BIS 3.0 BT-49: the buyer's electronic address (cbc:EndpointID) is
  // mandatory (1..1) — a buyer without a KVK number cannot receive a Peppol
  // document (the access point validates this). Fail early with a clear
  // message instead of a Schematron rejection at the provider.
  if (!invoice.contact?.kvk) {
    throw peppolError('PEPPOL_BUYER_MISSING_ID', `buyer '${invoice.contact?.name ?? 'unknown'}' has no KVK number — Peppol BIS 3.0 requires the buyer electronic address (BT-49); set it via 'contact update --id <id> --kvk <number>'`);
  }
  // Peppol BIS 3.0 BT-10 (PEPPOL-EN16931-R003, fatal; DE-R-015 for NL): the
  // buyer reference must be provided (there is no order-reference element in
  // this document). The invoice 'reference' field (klantkenmerk) is it.
  if (!invoice.reference) {
    throw peppolError('PEPPOL_BUYER_REFERENCE_MISSING', `invoice ${invoice.invoice_number} has no buyer reference — Peppol BIS 3.0 requires cbc:BuyerReference (BT-10); set it at creation with 'invoice create ... --reference <text>' or recreate the invoice`);
  }
  if (dryRun) {
    return {
      dryRun: true,
      invoice_number: invoice.invoice_number,
      endpoint: ep,
      configured: Boolean(cfg.token),
      bytes: invoiceToUbl(db, invoice).length,
    };
  }
  const xml = invoiceToUbl(db, invoice);
  let res;
  try {
    res = await fetch(ep, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/xml; charset=utf-8',
        ...(cfg.token ? { Authorization: `Bearer ${cfg.token}` } : {}),
      },
      body: xml,
      // never hang the CLI on a dead access point
      signal: AbortSignal.timeout(30_000),
    });
  } catch (err) {
    throw peppolError('PEPPOL_SEND_FAILED', `provider unreachable at ${ep}: ${err.message}`);
  }
  const text = await res.text();
  if (!res.ok) {
    throw peppolError('PEPPOL_SEND_FAILED', `provider returned ${res.status}: ${text.slice(0, 300)}`);
  }
  return {
    dryRun: false,
    invoice_number: invoice.invoice_number,
    endpoint: ep,
    status: res.status,
    response: text.slice(0, 500) || null,
  };
}

File: src/peppol.rs
// bukio-cli — agent-first double-entry bookkeeping for SMEs.
// Copyright (c) 2026 Erik van Kempen.
// SPDX-License-Identifier: Apache-2.0
//
// Peppol send (FR3.8): POST the UBL 2.1 / Peppol BIS 3.0 document to an
// access-point provider. The provider contract is deliberately thin: POST the
// XML with a bearer token, expect 2xx. Configuration comes from the
// environment — never hard-code credentials:
//   BUKIO_PEPPOL_ENDPOINT  e.g. https://api.storecove.com/api/v2/invoices
//   BUKIO_PEPPOL_TOKEN     the provider API token

use crate::money::{BukioError, Result};
use crate::ubl::invoice_to_ubl;
use rusqlite::Connection;
use serde_json::{json, Value};
use std::time::Duration;

fn peppol_error(code: &'static str, message: impl Into<String>) -> BukioError {
    BukioError::new(code, message)
}

/// (endpoint, token) from the environment.
pub fn peppol_config() -> (Option<String>, Option<String>) {
    (
        std::env::var("BUKIO_PEPPOL_ENDPOINT").ok(),
        std::env::var("BUKIO_PEPPOL_TOKEN").ok(),
    )
}

fn truncate(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}

/// Send a finalized invoice to the Peppol access point. dry_run validates the
/// configuration + payload without making the request.
pub fn send_peppol_invoice(
    db: &Connection,
    invoice: &Value,
    endpoint: Option<&str>,
    dry_run: bool,
) -> Result<Value> {
    let (cfg_endpoint, cfg_token) = peppol_config();
    let ep = match endpoint.or(cfg_endpoint.as_deref()) {
        Some(e) => e.to_string(),
        None => {
            return Err(peppol_error(
                "PEPPOL_NOT_CONFIGURED",
                "no Peppol endpoint — set BUKIO_PEPPOL_ENDPOINT (or pass --endpoint)",
            ))
        }
    };
    // Peppol BIS 3.0 BT-49: the buyer's electronic address (cbc:EndpointID) is
    // mandatory — a buyer without a KVK number cannot receive a Peppol
    // document. Fail early instead of taking a Schematron rejection later.
    if invoice["contact"]["kvk"].as_str().unwrap_or("").is_empty() {
        let name = invoice["contact"]["name"].as_str().unwrap_or("unknown");
        return Err(peppol_error(
            "PEPPOL_BUYER_MISSING_ID",
            format!("buyer '{name}' has no KVK number — Peppol BIS 3.0 requires the buyer electronic address (BT-49); set it via 'contact update --id <id> --kvk <number>'"),
        ));
    }
    // Peppol BIS 3.0 BT-10 (PEPPOL-EN16931-R003, fatal): the buyer reference is
    // required. The invoice `reference` field (klantkenmerk) is it.
    if invoice["reference"].as_str().unwrap_or("").is_empty() {
        return Err(peppol_error(
            "PEPPOL_BUYER_REFERENCE_MISSING",
            format!(
                "invoice {} has no buyer reference — Peppol BIS 3.0 requires cbc:BuyerReference (BT-10); set it at creation with 'invoice create ... --reference <text>' or recreate the invoice",
                invoice["invoice_number"].as_str().unwrap_or("")
            ),
        ));
    }
    let xml = invoice_to_ubl(db, invoice)?;
    let invoice_number = invoice["invoice_number"].clone();
    if dry_run {
        return Ok(json!({
            "dryRun": true,
            "invoice_number": invoice_number,
            "endpoint": ep,
            "configured": cfg_token.is_some(),
            "bytes": xml.chars().count(),
        }));
    }
    let config = ureq::config::Config::builder()
        .http_status_as_error(false)
        .timeout_global(Some(Duration::from_secs(30)))
        .build();
    let agent = ureq::Agent::new_with_config(config);
    let mut req = agent
        .post(&ep)
        .header("Content-Type", "application/xml; charset=utf-8");
    if let Some(ref tok) = cfg_token {
        req = req.header("Authorization", &format!("Bearer {tok}"));
    }
    let resp = req.send(xml.as_bytes()).map_err(|e| {
        peppol_error(
            "PEPPOL_SEND_FAILED",
            format!("provider unreachable at {ep}: {e}"),
        )
    })?;
    let status = resp.status().as_u16();
    let text = resp.into_body().read_to_string().map_err(|e| {
        peppol_error(
            "PEPPOL_SEND_FAILED",
            format!("failed to read provider response: {e}"),
        )
    })?;
    if !(200..300).contains(&status) {
        return Err(peppol_error(
            "PEPPOL_SEND_FAILED",
            format!("provider returned {status}: {}", truncate(&text, 300)),
        ));
    }
    Ok(json!({
        "dryRun": false,
        "invoice_number": invoice_number,
        "endpoint": ep,
        "status": status,
        "response": if text.is_empty() { Value::Null } else { json!(truncate(&text, 500)) },
    }))
}
